Implement incrprod

labiej opened the pull request:
Resolves #24 .

<!--lint disable first-heading-level-->

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

* [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
* [x] Read and understand the [Code of Conduct][code-of-conduct].
* [x] Read and understood the [licensing terms][license].
* [x] Searched for existing issues and pull requests __before__ submitting this pull request.
* [x] Filed an issue (or an issue already existed) __prior to__ submitting this pull request.
* [x] Rebased onto latest `develop`.
* [x] Submitted against `develop` branch.


## Description

> What is the purpose of this pull request?

This pull request:

* Provides an implementation for incrementally calculating a product

## Related Issues

> Does this pull request have any related issues?

This pull request:

* resolves #24 


## Questions

> Any questions for reviewers of this pull request?

* Are the tests extensive enough?
* Should a warning be added in the README.md related to the problems with resolvable finite products as encountered during testing?


## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.


---

@stdlib-js/reviewers


<!-- <links> -->

[contributing]: https://github.com/stdlib-js/stdlib/blob/master/CONTRIBUTING.md
[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/master/CODE_OF_CONDUCT.md
[license]: https://github.com/stdlib-js/stdlib/blob/master/LICENSE

<!-- </links> -->

kgryte committed:
Add revised Spache word list to dataset names

kgryte committed:
Add utility to uncurry a curried function

kgryte committed:
Add utility to uncurry a right curried function

kgryte committed:
Update namespace

labiej committed:
Add incrprod functionality, tests included

Minor changes to conform with other functionality

Fix minor differences to conform with style guide

Fix minor differences to conform with style guide again

Conform to style guide part 3

labiej committed:
Merge remote-tracking branch 'upstream/develop' into incrprod

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+// returns 2.0
+
+product = accumulator( 1.0 );
+// returns 3.0
```
This should be

``` javascript
// returns 2.0
```
correct?

labiej replied:
Indeed, fixed in the upcoming commit

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+## Notes
+
+* Input values are __not__ type checked. If provided `NaN` or a value which, when used in computations, results in `NaN`, the accumulated value is `NaN` for __all__ future invocations. If non-numeric inputs are possible, you are advised to type check and handle accordingly __before__ passing the value to the accumulator function.
+* For long running accumulations or accumulations of large numbers, care should be taken to prevent overflow. Overflow need not be forever since internally the accumulator stores the product as a __mantissa__ and __exponent__.
```
What happens in the case of underflow? Is recovering from underflow also possible given the algorithm?

labiej replied:
Recovering from underflow was possible in most cases before. The problem one could encounter was when a really long product occurred whence the mantissa could underflow.

Recovering from underflow is now always possible owing to the fact that the mantissa is normalized whenever it drops below 0.5.
Going below 0.5 is the only way the mantissa can leave the range [0.5, 1).

kgryte replied:
Thanks for the clarification. 👍

kgryte replied:
Should update the note regarding overflow to include underflow. e.g., 

> For long running accumulations or accumulations of either large or small numbers, care should be taken to prevent overflow and underflow. Note, however, that overflow/underflow may be transient, as the accumulator does not use a double to store an accumulated product, but instead, updates the normalized __mantissa__ and __exponent__ of a product separately. Doing so guards against a loss in precision.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+
+acc(z);
+console.log( "Multiplication with z removes the overflow\t" + acc() );
+
```
Remove trailing empty line.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+acc( y );
+console.log( "Multiplying x and y yields an overflow\t" + acc() );
+
+acc(z);
```
``` javascript
acc( z );
```

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+var ldexp = require('@stdlib/math/base/special/ldexp');
+var incrprod = require('@stdlib/math/generics/statistics/incrprod');
+
+/* Write huge numbers x and y as a double starting from a normalized mantissa and big exponent */
```
Comment style: `//`

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+var x = ldexp( 0.5, 1000 );
+var y = ldexp( 0.1, 1000 );
+
+/* The number z on the other hand is practically zero */
```
Comment style: `//`

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+
+var accumulator = incrprod();
+
+console.log( "x has a value\t" + x );
```
Single quotes here and throughout this example.

kgryte replied:
Rather than logging to the console here,

``` javascript
var x = ldexp( 0.5, 1000 );
// returns <value>

var y = ldexp( 0.1, 1000 );
// returns <value>

// ...

var out = acc( x );
// returns <value>

out = acc( y );
// returns <value>

out = acc( z );
// returns <value>
```
where `<value>` indicates the actual numeric value.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
````diff
@@ -0,0 +1,123 @@
+console.log( accumulator() );
+```
+
+### Overflow can vanish
````
Move to usage section and replace heading with

> Under certain conditions, overflow may be "corrected". For example,

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
````diff
@@ -0,0 +1,123 @@
+### Overflow can vanish
+
+``` javascript
+'use strict'
````
Remove `'use strict'` directive.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+'use strict'
+
+var ldexp = require('@stdlib/math/base/special/ldexp');
+var incrprod = require('@stdlib/math/generics/statistics/incrprod');
```
Pending the move to the "usage" section, can remove the `require` statement.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,123 @@
+
+## Examples
+
+### Default usage
```
Remove heading.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,46 @@
+'use strict'
+
+var incrprod = require( './../lib' );
```
This `require` statement should be the last require statement.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,46 @@
+console.log( '\nFinal product: %d\n', accumulator() );
+
+console.log( '\nTesting overflows\n\n');
+
```
Remove empty line.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,46 @@
+console.log( '\nTesting overflows\n\n');
+
+
+/* Write huge numbers x and y as a double starting from a normalized mantissa and big exponent */
```
Comment style: `//`

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,46 @@
+var x = ldexp( 0.5, 1000 );
+var y = ldexp( 0.1, 1000 );
+
+/* The number z on the other hand is practically zero */
```
Comment style: `//`

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,46 @@
+
+accumulator = incrprod();
+
+console.log( "x = " + x );
```
Same items as stated wrt README.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,57 @@
+var ldexp = require( '@stdlib/math/base/special/ldexp' );
+
+/**
+ * Returns an accumulator function which incrementally computes a product.
```
Comment alignment. No leading space. 

```
/**
*
*
*/
```

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,57 @@
+        if ( arguments.length === 0 ) {
+            return product;
+        }
+
```
Remove empty line.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        var newFactorized = frexp( x );
+        // Multiply mantissa
```
Empty line above comment.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,57 @@
+        var newFactorized = frexp( x );
+        // Multiply mantissa
+        productFactorized[ 0 ] *= newFactorized[ 0 ];
+        // Add exponents
```
Empty line above comment.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,57 @@
+        productFactorized[ 0 ] *= newFactorized[ 0 ];
+        // Add exponents
+        productFactorized[ 1 ] += newFactorized[ 1 ];
+
```
Should the fraction not be kept in normalized form? Here, given enough repetitions, the fraction may underflow, no? Keeping the fraction on the interval `[0.5, 1)` would require multiplying, using `frexp`, and then adding the "carried" exponent to the exponent computed at line `45`. Or maybe I am missing something?

labiej replied:
That's correct, I figured that out a couple hours after issuing the pull request.

It should be a trivial expansion.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,140 @@
+// MODULES //
+
+var tape = require( 'tape' );
+var incrprod = require( './../lib' );
```
This should be the last `require` statement.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,140 @@
+    var i;
+
+    acc = incrprod();
+    for( i = 0; i < length; i++ ) {
```
Missing space between `for` and `(`. If you run `make init` in your local repository, this should setup Git hooks for linting. Alternatively,

``` bash
$ make lint-javascript-src SOURCES_FILTER=.*/incrprod/.*
$ make lint-javascript-examples EXAMPLES_FILTER=.*/incrprod/.*
$ make lint-javascript-tests TESTS_FILTER=.*/incrprod/.*
```

labiej replied:
Thanks for the tip, I'll check if I can write a short guide from a novice's point of view.
It would be very simple but a base document could be useful regardless.

kgryte replied:
Awesome. Thanks!

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,140 @@
+
+    acc = incrprod();
+    for( i = 0; i < length; i++ ) {
+        if( i%2 == 0 ) {
```
Missing space.

kgryte replied:
Use strict equality: `===`.

labiej committed:
Keep mantissa normalized

Keeping the mantissa normalized also deals with underflow as long as the
new value passed to the accumulator is not causing underflow by itself

labiej committed:
Fix style issues

Final styling remark solved

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,149 @@
+
+## Examples
+
+
```
Remove extra empty line.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,149 @@
+var incrprod = require('@stdlib/math/generics/statistics/incrprod');
+
+// Write huge numbers x and y as a double starting from a normalized mantissa and big exponent
+
```
Remove empty line.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,149 @@
+// returns 1.0715086071862674e+300
+
+// The number z on the other hand is practically zero
+var z = ldexp( 0.2 , -1000 );
```
Remove extra space after `0.2`.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,149 @@
+
+var accumulator = incrprod();
+
+accumulator( x );
```
```
var prod = accumulator( x );
// returns ...

prod = accumulator ( y );
// returns ...

prod = accumulator( z );
// returns ...
```

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,149 @@
+// returns 9.332636185032189e-303
+
+// The number z on the other hand is very large
+var z = ldexp( 0.2 , 1000 );
```
Remove extra space after `0.2`.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+
+// Write huge numbers x and y as a double starting from a normalized mantissa and big exponent
+var x = ldexp( 0.5, 1000 );
+// Returns 5.357543035931337e+300
```
s/Returns/returns/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+var x = ldexp( 0.5, 1000 );
+// Returns 5.357543035931337e+300
+var y = ldexp( 0.1, 1000 );
+// Returns 1.0715086071862674e+300
```
s/Returns/returns/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+
+// The number z on the other hand is practically zero
+var z = ldexp( 0.2, -1000 );
+// Returns 1.8665272370064379e-302
```
s/Returns/returns/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+// Write huge numbers x and y as a double starting from a normalized mantissa and big exponent
+var x = ldexp( 0.5, 1000 );
+// Returns 5.357543035931337e+300
+var y = ldexp( 0.1, 1000 );
```
Insert empty line above.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+
+accumulator = incrprod();
+
+accumulator( x );
```
```
prod = accumulator( x );
...
```

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+console.log( '\nFinal product: %d\n', accumulator() );
+console.log( '\nTesting overflows\n\n');
+
+
```
Remove empty line.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,72 @@
+	* // returns -10.0
+	*/
+	function accumulator( x ) {
+		var newFactorized = frexp( x );
```
This is misplaced. If `x` is `undefined` (i.e., no arguments), we should not be calling `frexp`.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,72 @@
+		}
+
+		// Multiply mantissa
+		productFactorized[ 0 ] *= newFactorized[ 0 ];
```
What is the advantage of keeping the fraction and exponent stored in an `array`? May be a bit more clear to at `L17` or thereabout

``` javascript
var exponent;
var fraction;

// Initial product is 1.0, which may be split into its fractional and exponent parts (0.5 x 2.0**1 = 1.0):
fraction = 0.5;
exponent = 1.0;
```
Then at `L40`,

``` javascript
var parts;
if ( arguments.length === 0 ) {
    return ldexp( fraction, exponent );
}
// Split an incoming value into its fractional and exponent parts:
parts = frexp( x );

// Update the fraction:
fraction *= parts[ 0 ];

// Update the accumulated exponent:
exponent += parts[ 1 ];

// Ensure that the fraction is normalized to avoid overflow/underflow...
if ( fraction < 0.5 && fraction > -0.5 ) {
    // ...
}
// Return a double:
return ldexp( fraction, exponent );
```

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,72 @@
+
+		return product;
+	}
+} // End FUNCTION incrprod()
```
s/End/end/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/package.json:
```diff
@@ -0,0 +1,44 @@
+    "math",
+    "multiplication",
+    "product",
+    "total",
```
Add the following keywords:

* prod
* mult

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,143 @@
+	var actual;
+	var data;
+	var acc;
+	var prod;
```
Order by variable length, so switch order of `acc` and `prod`.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,143 @@
+	acc( x );
+	acc( y );
+
+	t.equal( acc(), Number.POSITIVE_INFINITY, 'return accumulated product 0.6*10^2000' );
```
Use `@stdlib/math/constants/float64-pinf` package rather than global.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,143 @@
+
+	t.equal( acc(), prod, 'the incremental product is correct' );
+	t.end();
+} );
```
Need to add tests for underflow similar to overflow. Additionally, need to check that, once provided an `x` which generates a `NaN`, a `NaN` is always returned (recovery is not possible).

labiej replied:
Should be resolved, any idea why it isn't shown as outdated?

kgryte replied:
Because, I believe, GitHub links a comment to a particular line. If a line associated with a review comment changes, then GitHub shows the comment as "outdated", even if the change does not resolve the issues raised in the comment. Here, the line associated with the comment did not change, even though the issue was resolved.

labiej committed:
Refactor code to avoid calls to unexisting arguments, store fraction and exponent separate

Add tests for underflow and NaN

labiej committed:
Resolve style issues, expand keywords in package.json

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,147 @@
+## Notes
+
+* Input values are __not__ type checked. If provided `NaN` or a value which, when used in computations, results in `NaN`, the accumulated value is `NaN` for __all__ future invocations. If non-numeric inputs are possible, you are advised to type check and handle accordingly __before__ passing the value to the accumulator function.
+* For long running accumulations or accumulations of either large or small numbers, care should be taken to prevent overflow and underflow. Note, however, that overflow/underflow may be transient, as the accumulator does not use a double to store an accumulated product, but instead, updates the normalized __mantissa__ and __exponent__ of a product separately. Doing so guards against a loss in pre ...
```
s/__mantissa__/__fraction__/ to ensure consistent language, including with `frexp`.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
````diff
@@ -0,0 +1,147 @@
+// returns 1.0715086071862675e+299
+```
+
+The output of the accumulator can underflow but this is recoverable. This becomes clear when considering that the actual multiplication uses the (normalized) mantissa and exponent. The following example shows how this works
````
Replace line with

> Under certain conditions, underflow may be "corrected". For example,

Would prefer to keep from "leaking" details of underlying implementation to a minimum. I sense that the sentence "when considering that the actual multiplication uses the..." will be confusing to most readers. Best to just remove in its entirety.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,147 @@
+var ldexp = require('@stdlib/math/base/special/ldexp');
+var incrprod = require('@stdlib/math/generics/statistics/incrprod');
+
+// Write numbers x and y, of tiny magnitude, as a double starting from a normalized mantissa and very negative exponent
```
Replace comment with

> // Create doubles of tiny magnitude:

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
```diff
@@ -0,0 +1,147 @@
+var ldexp = require('@stdlib/math/base/special/ldexp');
+var incrprod = require('@stdlib/math/generics/statistics/incrprod');
+
+// Write huge numbers x and y as a double starting from a normalized mantissa and big exponent
```
Replace comment with

> // Create doubles of huge magnitude:

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/examples/index.js:
```diff
@@ -0,0 +1,45 @@
+console.log( '\nFinal product: %d\n', accumulator() );
+console.log( '\nTesting overflows\n\n');
+
+// Write huge numbers x and y as a double starting from a normalized mantissa and big exponent
```
Include changes from README.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,79 @@
+			return product;
+		}
+
+		// Splitting incoming value into a normalised fraction and exponent
```
Consistent spelling (American).

s/normalised/normalized/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,79 @@
+		// Update the exponent
+		exponent += parts[ 1 ];
+
+		// Ensure fraction remains normalised to avoid overflow/underflow
```
Consistent spelling (American):

s/normalised/normalized/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,79 @@
+
+		// Ensure fraction remains normalised to avoid overflow/underflow
+		if ( fraction < 0.5 && fraction > -0.5) {
+			// Split fraction into a normalised fraction and exponent
```
Consistent spelling (American):

s/normalised/normalized/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/lib/incrprod.js:
```diff
@@ -0,0 +1,79 @@
+		// Store resulting double for future calls
+		product = ldexp( fraction, exponent );
+
+		// Seturn a double and store it for future calls
```
Remove comment, as you have already stated that the `product` is stored for future calls above and the `return` is self-evident.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/package.json:
```diff
@@ -0,0 +1,46 @@
+{
+  "name": "@stdlib/math/generics/statistics/incrsum",
```
s/incrsum/incrprod/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/package.json:
```diff
@@ -0,0 +1,46 @@
+    "mathematics",
+    "math",
+    "multiplication",
+		"mult",
```
Spacing.

labiej replied:
This one was annoying, atom showed the correct spacing but git(hub) knew better. Opening it in another editor fixed it.

kgryte replied:
Do you use the `editorconfig` plugin? This, in addition to ESLint/JSHint, is used to check spacing. In the future, at least for `package.json`, this won't be an issue once I add a tool to auto format `package.json` files. Should happen within the next couple weeks.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/package.json:
```diff
@@ -0,0 +1,46 @@
+    "multiplication",
+		"mult",
+    "product",
+		"prod",
```
Spacing.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,195 @@
+// MODULES //
+
+var tape = require( 'tape' );
+var POS_INFINITY = require( '@stdlib/math/constants/float64-pinf' );
```
Common practice within the project is to alias positive infinity as `PINF`.

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
````diff
@@ -0,0 +1,147 @@
+
+The output of the accumulator can underflow but this is recoverable. This becomes clear when considering that the actual multiplication uses the (normalized) mantissa and exponent. The following example shows how this works
+
+```javascript
````
Space before `javascript`.

kgryte reviewed (changes requested):
Major issues are resolved. Only minor nits remain. Thanks for your work on this!

labiej committed:
Normalize language and clarify comments

labiej committed:
Add example about underflows and fix typo in README.md

Change spacing of mult and prod keywords

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/README.md:
````diff
@@ -0,0 +1,147 @@
+// returns 1.0715086071862675e+299
+```
+
+Under certain conditions, overflow may be "corrected". For example,
````
s/overflow/underflow/

kgryte commented on lib/node_modules/@stdlib/math/generics/statistics/incrprod/test/test.js:
```diff
@@ -0,0 +1,195 @@
+
+	acc( x );
+
+	t.equal( typeof acc() === 'number', true, 'accumulator works correctly' );
```
This statement can be simplified to

```
t.equal( typeof acc(), 'number', 'returns a numeric value' );
```

kgryte reviewed (changes requested):
Typo and a simplification.

labiej committed:
Fix typo and simplify test of NaN persistence

kgryte commented:
@labiej Closing, as I did a manual merge by cherry picking the relevant commits in order to avoid history conflicts. Thanks for the addition!

Closes #24.
